src/corehost/src/tpafile.cpp

```diff
@@ -57,195 +75,188 @@ bool read_field(pal::string_t line, int& offset, pal::string_t& value_recv)
+
+            pal::string_t file_path = dir + DIR_SEPARATOR + file;            
+            trace::verbose(_X("adding %s to local list from %s"), file_name.c_str(), file_path.c_str());
+            m_local_assemblies.emplace(file_name, file_path);
         }
     }
 }
 
-void tpafile::write_tpa_list(pal::string_t& output)
+void add_tpa_asset(
+    const pal::string_t& asset_name,
+    const pal::string_t& asset_path,
+    std::set<pal::string_t>* items,
+    pal::string_t* output)
 {
+    if (items->count(asset_name))
+    {
+        return;
+    }
+
+    trace::verbose(_X("adding tpa entry: %s"), asset_path.c_str());
+    output->append(asset_path);
+    output->push_back(PATH_SEPARATOR);
+    items->insert(asset_name);
+}
+
+void tpafile::write_tpa_list(
+        const pal::string_t& app_dir,
+        const pal::string_t& package_dir,
+        const pal::string_t& clr_dir,
+        pal::string_t& output)
+{
+    get_local_assemblies(app_dir);
+
     std::set<pal::string_t> items;
-    for (auto entry : m_entries)
+    for (const tpaentry_t& entry : m_tpa_entries)
     {
-        if (pal::strcmp(entry.asset_type.c_str(), _X("runtime")) == 0 && items.find(entry.asset_name) == items.end())
+        if (entry.asset_type != _X("runtime") || items.count(entry.asset_name))
         {
-            // Resolve the full path
-            for (auto search_path : m_package_search_paths)
-            {
-                pal::string_t candidate;
-                candidate.reserve(search_path.length() +
-                    entry.library_name.length() +
-                    entry.library_version.length() +
-                    entry.relative_path.length() + 3);
-                candidate.append(search_path);
-
-                append_path(candidate, entry.library_name.c_str());
-                append_path(candidate, entry.library_version.c_str());
-                append_path(candidate, entry.relative_path.c_str());
-                if (pal::file_exists(candidate))
-                {
-                    trace::verbose(_X("adding tpa entry: %s"), candidate.c_str());
-
-                    output.append(candidate);
-                    output.push_back(PATH_SEPARATOR);
-                    items.insert(entry.asset_name);
-                    break;
-                }
-            }
+            continue;
+        }
+
+        pal::string_t redirection_path, candidate;
+        if (entry.library_type == _X("Package") &&
+                m_svc->find_redirection(entry.library_name, entry.library_version, entry.relative_path, &redirection_path))
+        {
+            add_tpa_asset(entry.asset_name, redirection_path, &items, &output);
+        }
+        else if (m_local_assemblies.count(entry.asset_name))
+        {
+            // TODO: Case insensitive look up?
+            add_tpa_asset(entry.asset_name, m_local_assemblies.find(entry.asset_name)->second, &items, &output);
+        }
+        else if (entry.to_full_path(package_dir, &candidate))
```

Comment:
After AppLocal, we should also look at DOTNET_PACKAGES_CACHE (used in Antares) and then look at package_dir (which maps to DOTNET_PACKAGES) location.